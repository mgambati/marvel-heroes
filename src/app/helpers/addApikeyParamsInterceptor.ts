import { AxiosRequestConfig } from "axios";

/**
 * Devido a um bug no Axios, default params n√£o sofre merge com params da request;
 * Para resolver isso usa-se um interceptor
 */
export function addApikeyParamsInterceptor(
  config: AxiosRequestConfig
): AxiosRequestConfig {
  config.params = {
    ...config.params,
    apikey: process.env["REACT_APP_MARVEL_API_KEY"],
  };

  return config;
}
