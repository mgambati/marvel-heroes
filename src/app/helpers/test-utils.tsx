import { render as rtlRender } from "@testing-library/react";
import React, { FC } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider, Theme } from "theme-ui";
import { ThunkDispatch, AnyAction } from "@reduxjs/toolkit";
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import { RootState, store } from "../store";
import { BASE_THEME } from "../constants/BASE_THEME";

export function setupMockedStore() {
  type DispatchExts = ThunkDispatch<RootState, void, AnyAction>;
  const middlewares = [thunk];
  const mockStore = configureStore<RootState, DispatchExts>(middlewares);
  const initialState: RootState = {};

  return [mockStore, initialState] as [typeof mockStore, typeof initialState];
}

function render(
  ui: JSX.Element,
  { initialState = undefined, mockedStore = store, ...renderOptions } = {}
) {
  const Wrapper: FC = ({ children }) => {
    return (
      <ThemeProvider theme={BASE_THEME as Theme}>
        <BrowserRouter>
          <Provider store={store}>{children}</Provider>
        </BrowserRouter>
      </ThemeProvider>
    );
  };

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from "@testing-library/react";

export { render };
