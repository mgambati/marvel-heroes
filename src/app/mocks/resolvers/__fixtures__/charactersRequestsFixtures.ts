export const CHARACTERS_RESPONSE_PAYLOAD = {
  offset: 0,
  limit: 2,
  total: 4,
  count: 2,
  results: [
    {
      id: 1,
      name: "3-D Man",
      description: "",
      thumbnail: {
        path: "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
        extension: "jpg",
      },
    },
    {
      id: 2,
      name: "A-Bomb (HAS)",
      description:
        "Rick Jones has been Hulk's best bud since day one, but now he's more than a friend...he's a teammate! Transformed by a Gamma energy explosion, A-Bomb's thick, armored skin is just as strong and powerful as it is blue. And when he curls into action, he uses it like a giant bowling ball of destruction! ",
      thumbnail: {
        path: "http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16",
        extension: "jpg",
      },
    },
  ],
};

export const EMPTY_CHARACTERS_RESPONSE_PAYLOAD = {
  offset: 2,
  limit: 2,
  total: 4,
  count: 0,
  results: [],
};

export const SINGLE_CHARACTERS_RESPONSE_PAYLOAD = {
  id: 1,
  name: "3-D Man",
  description: "",
  modified: "2014-04-29T14:18:17-0400",
  thumbnail: {
    path: "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
    extension: "jpg",
  },
};

export const STARTS_WITH_CHARACTERS_RESPONSE_PAYLOAD = {
  offset: 0,
  limit: 1,
  total: 2,
  count: 1,
  results: [
    {
      id: 3,
      name: "Loki",
      description: "",
      modified: "2017-08-21T16:45:34-0400",
      thumbnail: {
        path: "http://i.annihil.us/u/prod/marvel/i/mg/d/90/526547f509313",
        extension: "jpg",
      },
    },
  ],
};
