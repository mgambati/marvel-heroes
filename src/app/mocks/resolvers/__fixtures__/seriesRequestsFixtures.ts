export const SERIES_BY_CHARACTER_RESPONSE_PAYLOAD = {
  offset: 0,
  limit: 2,
  total: 3,
  count: 2,
  results: [
    {
      id: 1945,
      title: "Avengers: The Initiative (2007 - 2010)",
      description: undefined,
      thumbnail: {
        path: "http://i.annihil.us/u/prod/marvel/i/mg/5/a0/514a2ed3302f5",
        extension: "jpg",
      },
    },
    {
      id: 2005,
      title: "Deadpool (1997 - 2002)",
      description:
        "Wade Wilson: Heartless Merc With a Mouth or...hero? Laugh, cry and applaud at full volume for the mind-bending adventures of Deadpool, exploring the psyche and crazed adventures of Marvel's most unstable personality!",
      thumbnail: {
        path: "http://i.annihil.us/u/prod/marvel/i/mg/7/03/5130f646465e3",
        extension: "jpg",
      },
    },
  ],
};
