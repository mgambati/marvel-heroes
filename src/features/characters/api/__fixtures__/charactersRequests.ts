export const CHARACTERS_RESPONSE_PAYLOAD = {
  offset: 0,
  limit: 2,
  total: 4,
  count: 2,
  results: [
    {
      id: 1011334,
      name: "3-D Man",
      description: "",
      thumbnail: {
        path: "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
        extension: "jpg",
      },
    },
    {
      id: 1017100,
      name: "A-Bomb (HAS)",
      description:
        "Rick Jones has been Hulk's best bud since day one, but now he's more than a friend...he's a teammate! Transformed by a Gamma energy explosion, A-Bomb's thick, armored skin is just as strong and powerful as it is blue. And when he curls into action, he uses it like a giant bowling ball of destruction! ",
      thumbnail: {
        path: "http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16",
        extension: "jpg",
      },
    },
  ],
};

export const EMPTY_CHARACTERS_RESPONSE_PAYLOAD = {
  offset: 2,
  limit: 2,
  total: 4,
  count: 0,
  results: [],
};

export const SINGLE_CHARACTERS_RESPONSE_PAYLOAD = {
  id: 1011334,
  name: "3-D Man",
  description: "",
  modified: "2014-04-29T14:18:17-0400",
  thumbnail: {
    path: "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
    extension: "jpg",
  },
};
