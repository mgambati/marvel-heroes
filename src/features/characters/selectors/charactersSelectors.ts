import { RootState } from "../../../app/store";
import { charactersAdapter } from "../slices/charactersSlice";
import { createSelector } from "@reduxjs/toolkit";

const charactersSelectors = charactersAdapter.getSelectors();

const selectCharacterState = (state: RootState) => state.characters;

export const selectAllCharactersIds = createSelector(
  selectCharacterState,
  charactersSelectors.selectIds
);

export const createSelectCharacterById = (id: number | string) => {
  return (state: RootState) =>
    charactersSelectors.selectById(state.characters, id);
};

export const createSelectStatusOfLoadCharacterById = (id: number | string) => {
  return (state: RootState) => state.characters.status[id];
};

export const selectStatusOfLoadCharacters = (state: RootState) =>
  state.characters.status["all"];

export const selectCurrentCharactersQuery = (state: RootState) =>
  state.characters.query;

export const selectCurrentCharactersOffset = (state: RootState) =>
  state.characters.offset;
