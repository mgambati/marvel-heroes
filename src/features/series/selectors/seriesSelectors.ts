import { seriesAdapter } from "../slices/seriesSlice";
import { RootState } from "../../../app/store";

const seriesSelectors = seriesAdapter.getSelectors();

export const createSelectSeriesByCharacter = (characterId: number | string) => {
  return (state: RootState) => state.series.byCharacter[characterId];
};

export const createSelectSeriesById = (id: number | string) => {
  return (state: RootState) => seriesSelectors.selectById(state.series, id);
};

export const createSelectStatusOfLoadSeriesByCharacter = (
  characterId: number | string
) => {
  return (state: RootState) => state.series.status[`character.${characterId}`];
};
