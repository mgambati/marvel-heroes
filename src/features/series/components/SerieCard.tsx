import React, { FC, useMemo } from "react";
import { useSelector } from "react-redux";
import { AspectRatio, Card, Flex, Image, Text } from "theme-ui";
import { createSelectSeriesById } from "../selectors/seriesSelectors";

export interface SeriesCardProps {
  id: number | string;
}

const SerieCard: FC<SeriesCardProps> = ({ id }) => {
  const selectSeriesById = useMemo(() => createSelectSeriesById(id), [id]);

  const series = useSelector(selectSeriesById);
  const thumbnailSource = `${series?.thumbnail.path}.${series?.thumbnail.extension}`;

  return (
    <Card
      sx={{
        display: "flex",
        borderRadius: "lg",
        flexDirection: "column",
        overflow: "hidden",
        bg: "red.7",
        ":hover": { opacity: 0.8 },
      }}
      data-testid="serie-card"
    >
      <AspectRatio ratio={1}>
        <Image
          src={thumbnailSource}
          sx={{
            width: "100%",
            objectFit: "cover",
            objectPosition: "center",
            height: "100%",
          }}
        />
      </AspectRatio>
      <Flex
        sx={{
          flexGrow: 1,
          justifyContent: "center",
          flexDirection: "column",
        }}
      >
        <Text sx={{ m: 2 }} color="white">
          {series?.title}
        </Text>
      </Flex>
    </Card>
  );
};

export default SerieCard;
