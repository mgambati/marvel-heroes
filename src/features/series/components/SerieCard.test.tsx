import React from "react";
import { render } from "../../../app/helpers/test-utils";
import { SINGLE_CHARACTERS_RESPONSE_PAYLOAD } from "../../../app/mocks/resolvers/__fixtures__/charactersRequestsFixtures";
import { SERIES_BY_CHARACTER_RESPONSE_PAYLOAD } from "../../../app/mocks/resolvers/__fixtures__/seriesRequestsFixtures";
import * as charactersSelectors from "../selectors/seriesSelectors";
import SerieCard from "./SerieCard";

test("renders correctly", () => {
  const serie = SERIES_BY_CHARACTER_RESPONSE_PAYLOAD.results[0];
  jest
    .spyOn(charactersSelectors, "createSelectSeriesById")
    .mockReturnValue(() => serie);

  const { container } = render(
    <SerieCard id={SINGLE_CHARACTERS_RESPONSE_PAYLOAD.id} />
  );

  expect(container).toMatchSnapshot();
});
