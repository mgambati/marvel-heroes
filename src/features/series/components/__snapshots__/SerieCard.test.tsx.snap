// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    class="css-y04738"
    data-testid="serie-card"
  >
    <div
      class="css-i9l3dq"
    >
      <div
        class="css-tmeqbf"
      />
      <div
        class="css-ojfau8"
      >
        <img
          class="css-dqm8hw"
          src="http://i.annihil.us/u/prod/marvel/i/mg/5/a0/514a2ed3302f5.jpg"
        />
      </div>
    </div>
    <div
      class="css-seap5o"
    >
      <div
        class="css-ghx28c"
      >
        Avengers: The Initiative (2007 - 2010)
      </div>
    </div>
  </div>
</div>
`;
